datasquare.go

```diff
@@ -176,8 +176,15 @@ func (ds *dataSquare) setColSlice(x uint, y uint, newCol [][]byte) error {
 }
 
 func (ds *dataSquare) resetRoots() {
-	ds.rowRoots = nil
-	ds.colRoots = nil
+	// don't write nil if it's already nil
+	// this prevents rewriting nil into shared memory slot
+	// when resetRoots is used from multiple routines
+	if ds.rowRoots != nil {
```

Comment:
Instead of this, could we use a mutex? E.g. functions other than `SetCell` use a mutex prior to calling `resetRoots`

https://github.com/celestiaorg/rsmt2d/blob/ed5b81ebc9a0262ea08834a50ef05501bc8b74fc/datasquare.go#L165-L166

Reply:
The mutex is blocking. In my case there are hundreds of routines resetting the roots by calling SetCell. Contention between those should be avoided. The data structure in the nose around the EDs is non-blocking, and this solution keeps it so.

Orthogonally to multithreading, I feel like rewriting nil all the time is not a good thing to either, but I couldn't find solid proof for that.

Reply:
Right now the EDS does not have any guarantees regarding thread safety for reads and writes. If we introduce it, we should do it not just in one place but more holistically. The parallelization is internal for the API and orthogonal. 

Additionally, we can extract the SyncEDS from the node, so the lib has a synchronized yet nonblocking version of the square.